kmsjwt.go

```diff
@@ -5,105 +5,97 @@ import (
 	"crypto/rsa"
 	"crypto/x509"
 	"encoding/base64"
-	"sync"
+	"fmt"
 
 	"github.com/aws/aws-sdk-go-v2/aws"
 	"github.com/aws/aws-sdk-go-v2/service/kms"
 	"github.com/aws/aws-sdk-go-v2/service/kms/types"
 	"github.com/golang-jwt/jwt/v4"
-	"github.com/pkg/errors"
 )
 
-var signingMethod = jwt.SigningMethodRS512
+var (
+	signingMethod = jwt.SigningMethodPS512
+	_             = jwt.SigningMethod(KMSJWT{})
+)
 
-// KMSJWT is a JWT signing method implementation using RSA512 with the private
-// key stored in AWS KMS. The public key is retrieved from KMS on
-// initialization.
+// KMSJWT is a JWT signing method implementation using an asymmetric AWS KMS key.
+// The signing is done by KMS service, so there is a network call on every sign action.
+// The verification is done on the client side with the exported public key.
+// The public key is retrieved from KMS on initialization.
 type KMSJWT struct {
-	api   KMS
-	keyID string
-
-	lock      sync.Mutex
+	client    KMS
+	keyID     string
 	publicKey *rsa.PublicKey
 }
 
-// New provides a KMS-based implementation of JWT signing method.
-func New(api KMS, keyID string) *KMSJWT {
-	return &KMSJWT{api: api, keyID: keyID}
+// New retrieves the public key from KMS and returns a signer.
+func New(ctx context.Context, client KMS, keyID string) (KMSJWT, error) {
+	publicKey, err := getPublicKey(ctx, client, keyID)
+	if err != nil {
+		return KMSJWT{}, fmt.Errorf("kmsjwt new: %w", err)
+	}
+	return KMSJWT{client: client, keyID: keyID, publicKey: publicKey}, err
 }
 
-// NewWithPublicKey provides a KMS-based implementation of JWT signing method
-// with a pre-loaded public key.
-func NewWithPublicKey(api KMS, keyID string, publicKey *rsa.PublicKey) *KMSJWT {
-	return &KMSJWT{api: api, keyID: keyID, publicKey: publicKey}
+func getPublicKey(ctx context.Context, client KMS, keyID string) (*rsa.PublicKey, error) {
+	response, err := client.GetPublicKey(ctx, &kms.GetPublicKeyInput{KeyId: &keyID})
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve public key: %w", err)
+	}
+
+	publicKey, err := x509.ParsePKIXPublicKey(response.PublicKey)
+	if err != nil {
+		return nil, fmt.Errorf("could not parse public key: %w", err)
+	}
+
+	result, ok := publicKey.(*rsa.PublicKey)
+	if !ok {
+		return nil, fmt.Errorf("public key: cannot assert %T as %T", publicKey, result)
+	}
+
+	return result, nil
 }
 
-func (k *KMSJWT) Alg() string {
+// Alg returns the signing algorithm as defined in https://datatracker.ietf.org/doc/html/rfc7518#section-3.1.
+func (k KMSJWT) Alg() string {
```

Comment:
Any reason not to use pointer receivers?

```suggestion
func (k *KMSJWT) Alg() string {
```

Reply:
I usually approach this form the other direction. Any reason to use pointers?

Since the type does not need to modify the struct and the struct is small I defaulted to value receiver.

> For types such as basic types, slices, and small structs, a value receiver is very cheap so unless the semantics of the method requires a pointer, a value receiver is efficient and clear.

https://go.dev/doc/faq#methods_on_values_or_pointers

Reply:
I just thought we should avoid a copy of the pubkey. This struct does not look like a "basic" type to me, but that's a bit subjective.

Not a strong opinion, though. 

I'm usually more on the other side of the road, as it's a safer default in terms of memory usage. That's easy to forgot to switch to pointer receiver if we add more heavy fields to it.

Reply:
The `rsa.PublicKey` is actually a pointer, so small in the memory. Maybe it is somewhat strange to keep a pointer in a non-pointer struct. 🤔

I am also thinking now about extendability of the struct without breaking the interface. Because of that maybe it would worth the use of pointers. But I think we should be ok for now.

Reply:
> The rsa.PublicKey is actually a pointer

Ha indeed that's an interface 👍🏻 So I agree that's not a big deal, we could probably move on!

Reply:
We don't have a user base, I think (counting stars and forks is easy 😆). So I am fine to err on the value type side.